barretenberg/cpp/src/barretenberg/stdlib/plonk_recursion/aggregation_state/aggregation_state.hpp

```diff
@@ -92,18 +83,133 @@ template <typename Curve> void write(std::vector<uint8_t>& buf, aggregation_stat
 
     write(buf, as.P0);
     write(buf, as.P1);
-    write(buf, as.public_inputs);
-    write(buf, as.proof_witness_indices);
     write(buf, as.has_data);
 };
 
 template <typename NCT> std::ostream& operator<<(std::ostream& os, aggregation_state<NCT> const& as)
 {
     return os << "P0: " << as.P0 << "\n"
               << "P1: " << as.P1 << "\n"
-              << "public_inputs: " << as.public_inputs << "\n"
-              << "proof_witness_indices: " << as.proof_witness_indices << "\n"
               << "has_data: " << as.has_data << "\n";
 }
 
+/**
+ * @brief Converts a list of 16 witness indices corresponding to the bigfield limbs of an aggregation object to an
+ * aggregation_state.
+ *
+ * @tparam Builder
+ * @tparam Curve
+ * @param builder
+ * @param witness_indices
+ * @return aggregation_state<Curve>
+ */
+template <typename Builder, typename Curve>
+aggregation_state<Curve> convert_witness_indices_to_agg_obj(Builder& builder,
+                                                            AggregationObjectIndices const& witness_indices)
+{
+    aggregation_state<Curve> agg_obj;
+
+    size_t idx = 0;
+    std::array<typename Curve::Group, 2> pairing_points;
+    for (size_t i = 0; i < 2; i++) {
+        std::array<typename Curve::BaseField, 2> base_field_vals;
+        for (size_t j = 0; j < 2; j++) {
+            std::array<typename Curve::ScalarField, 4> bigfield_limbs;
+            for (size_t k = 0; k < 4; k++) {
+                bigfield_limbs[k] = Curve::ScalarField::from_witness_index(&builder, witness_indices[idx]);
```

Comment:
It feels like some additional checks should be necessary here, e.g. range constraints on the limbs like [here](https://github.com/AztecProtocol/aztec-packages/blob/d641a4bcce91c785e84cdfccb57a8ae120477e39/barretenberg/cpp/src/barretenberg/stdlib/plonk_recursion/verifier/verifier.hpp#L344)

Reply:
yeah, good point